yarn-project/simulator/src/public/public_tx_simulator.ts

```diff
@@ -0,0 +1,263 @@
+
+  constructor(
+    private db: MerkleTreeReadOperations,
+    private publicKernelSimulator: PublicKernelCircuitSimulator,
+    private globalVariables: GlobalVariables,
+    private worldStateDB: WorldStateDB,
+    private enqueuedCallSimulator: EnqueuedCallSimulator,
+    private publicKernelTailSimulator: PublicKernelTailSimulator,
+  ) {
+    this.log = createDebugLogger(`aztec:sequencer`);
+  }
+
+  static create(
+    db: MerkleTreeReadOperations,
+    publicExecutor: PublicExecutor,
+    publicKernelSimulator: PublicKernelCircuitSimulator,
+    globalVariables: GlobalVariables,
+    historicalHeader: Header,
+    worldStateDB: WorldStateDB,
+    realAvmProvingRequests: boolean = true,
+  ) {
+    const enqueuedCallSimulator = new EnqueuedCallSimulator(
+      db,
+      worldStateDB,
+      publicExecutor,
+      globalVariables,
+      historicalHeader,
+      realAvmProvingRequests,
+    );
+
+    const publicKernelTailSimulator = PublicKernelTailSimulator.create(db, publicKernelSimulator);
+
+    return new PublicTxSimulator(
+      db,
+      publicKernelSimulator,
+      globalVariables,
+      worldStateDB,
+      enqueuedCallSimulator,
+      publicKernelTailSimulator,
+    );
+  }
+
+  async process(tx: Tx): Promise<PublicTxResult> {
+    this.log.verbose(`Processing tx ${tx.getTxHash()}`);
+
+    const context = await PublicTxContext.create(this.db, this.worldStateDB, tx, this.globalVariables);
+
+    const setupResult = await this.processSetupPhase(context);
+    const appLogicResult = await this.processAppLogicPhase(context);
+    const teardownResult = await this.processTeardownPhase(context);
+
+    const processedPhases = [setupResult, appLogicResult, teardownResult].filter(
+      result => result !== undefined,
+    ) as ProcessedPhase[];
+
+    const _endStateReference = await this.db.getStateReference();
+    const transactionFee = context.getTransactionFee();
+
+    const tailKernelOutput = await this.publicKernelTailSimulator.simulate(context.latestPublicKernelOutput);
+
+    context.avmProvingRequest!.inputs.output = generateAvmCircuitPublicInputs(
+      tx,
+      tailKernelOutput,
+      context.getGasUsedForFee(),
+      transactionFee,
+    );
+
+    const gasUsed = {
+      totalGas: context.getActualGasUsed(),
+      teardownGas: context.teardownGasUsed,
+    };
+    return {
+      avmProvingRequest: context.avmProvingRequest!,
+      gasUsed,
+      revertCode: context.revertCode,
+      revertReason: context.revertReason,
+      processedPhases: processedPhases,
+    };
+  }
+
+  private async processSetupPhase(context: PublicTxContext): Promise<ProcessedPhase | undefined> {
+    // Start in phase TxExecutionPhase.SETUP;
+    if (context.hasPhase()) {
+      return await this.processPhase(context);
+    }
+  }
+
+  private async processAppLogicPhase(context: PublicTxContext): Promise<ProcessedPhase | undefined> {
+    context.progressToNextPhase(); // to app logic
+    if (context.hasPhase()) {
```

Comment:
Same here, in principle you can't be sure (from the PoV of this function) that the context phase is indeed AppLogic.